src/python/pants/python/setup_py_runner.py

```diff
@@ -0,0 +1,151 @@
+                help="The wheel version to use when executing `setup.py` scripts.",
+            )
+
+        @classmethod
+        def create(
+            cls,
+            *,
+            pex_file_path: Optional[Path] = None,
+            extra_reqs: Optional[List[PythonRequirement]] = None,
+            interpreter: Optional[PythonInterpreter] = None,
+            scope: Optional[Optionable] = None,
+        ) -> "SetupPyRunner":
+            factory = cls.scoped_instance(scope) if scope is not None else cls.global_instance()
+            requirements_pex = factory.bootstrap(
+                interpreter=interpreter,
+                pex_file_path=pex_file_path or os.path.join(safe_mkdtemp(), "setup-py-runner.pex"),
+                extra_reqs=extra_reqs,
+            )
+            return SetupPyRunner(requirements_pex=requirements_pex)
+
+        @property
+        def base_requirements(self):
+            return [
+                PythonRequirement(f"setuptools=={self.get_options().setuptools_version}"),
+                PythonRequirement(f"wheel=={self.get_options().wheel_version}"),
+            ]
+
+    class CommandFailure(Exception):
+        """Indicates an error executing setup.py commands."""
+
+    def __init__(self, requirements_pex: PEX) -> None:
+        self._requirements_pex = requirements_pex
+
+    @memoized_method
+    def __str__(self) -> str:
+        pex_path = self._requirements_pex.path()
+        pex_info = PexInfo.from_pex(pex_path)
+        requirements = "\n  ".join(map(str, pex_info.requirements))
+        return f"{type(self).__name__} at {pex_path} with requirements:\n  {requirements} "
+
+    def _create_python_args(self, setup_command: Iterable[str]) -> Iterable[str]:
+        args = ["setup.py", "--no-user-cfg"]
+        args.extend(setup_command)
+        return args
+
+    def cmdline(self, setup_command: Iterable[str]) -> Iterable[str]:
+        """Returns the command line that would be used to execute the given setup.py command."""
+        args = self._create_python_args(setup_command)
+        cmdline: List[str] = self._requirements_pex.cmdline(args)
+        return cmdline
+
+    def run_setup_command(
+        self, *, source_dir: Path, setup_command: Iterable[str], **kwargs
+    ) -> None:
+        """Runs the given setup.py command against the setup.py project in `source_dir`.
+
+        :raises: :class:`SetupPyRunner.CommandFailure` if there was a problem executing the command.
+        """
+        with pushd(str(source_dir)):
+            result = self._requirements_pex.run(
+                args=self._create_python_args(setup_command), **kwargs
+            )
+            if result != 0:
+                pex_command = safe_shlex_join(self.cmdline(setup_command))
+                raise self.CommandFailure(f"Failed to execute {pex_command} using {self}")
+
+    def _collect_distribution(
+        self, source_dir: Path, setup_command: Iterable[str], dist_dir: Path
```

Comment:
Looks like these indeed must be `Path`, rather than `PurePath`, because you call `source_dir.is_dir()`.